docs/architecture/adr-069-flexible-node-intitalization.md

```diff
@@ -0,0 +1,253 @@
+possibility of this alternative.
+
+### Abstract Dependency Based Startup and Shutdown
+
+The proposal on in this document simplifies and makes tendermint node
+initialization more abstract, but the system lacks a number of
+features which daemon/service initialization might provide, such as a
+dependency based system that allows the authors of services to control
+the initialization and shutdown order of components using an
+dependencies to control the ordering.
+
+Dependency based orderings make it possible to write components
+(reactors, etc.) with only limited awareness of other components or of
+node initialization, and is the state of the art for process
+initialization. However, this may be too abstract and complicated, and
+authors of components in the current implementation of tendermint
+*would* need to know about other components, so a dependency based
+system would be unhelpfully abstract at this stage.
+
+## Decisions
+
+- Provide a more flexible internal framework for initializing tendermint
+  nodes to make the initatilization process less hard-coded by the
+  implementation of the node objects.
+
+  - Reactors should not need to expose their interfaces *within* the
+	implementation of the node type, except in the context of some groups of
+	services.
+
+  - This refactoring should be entirely opaque to users.
+
+  - These node initialization changes should not require a
+	reevaluation of the `service.Service` or a generic initialization
+	orchestration framework.
+
+- If required for the 0.35 release, add an "unsafe" way to construct a
+  node service with user-supplied mempool, including making the
+  relevant interfaces and types external/exported.
+
+- Prioritize implementation of p2p-based statesync reactor to obviate
+  need for users to inject a custom state-sync provider.
+
+## Detailed Design
+
+The [current
+nodeImpl](https://github.com/tendermint/tendermint/blob/master/node/node.go#L47)
+includes direct references to the implementations of each of the
+reactors, which should be replaced by references to `service.Service`
+objects. This will require moving the construction of the [rpc
+service](https://github.com/tendermint/tendermint/blob/master/node/node.go#L771)
+into the constructor of
+[makeNode](https://github.com/tendermint/tendermint/blob/master/node/node.go#L126).
+
+In order to prevent adding complexity to the `node` package, this
+project will add an implementation to the service `service` package
+that implements `service.Service` and is comprised of a sequence of
+underlying `service.Service` objects and handles their
+startup/shutdown in a consistent order.
+
+Consensus, blocksync (nee fast sync.), and statesync all depend on
+each other, and have significant initialization dependencies that are
+presently encoded in the `node` package, and as part of this change,
+a new package/component will encapsulate the initialization of this
+area of functionality.
```

Comment:
Mind describing what you had in mind for this? I think some kind of 'BlockOrchestrator' object makes sense here, but I'm wondering if you have ideas for what its responsibilities will be.

Reply:
I'm mostly imagining that there'll be a new "virtual service" in `internal/blocks` that will encapsulate this dependency cycle/system, and I think could eventually fulfill the role of "block orchestrator".